File: usm-api/src/main/java/com/github/zly2006/usm/api/instance/Instance.java
package com.github.zly2006.usm.api.instance;

import com.github.zly2006.usm.api.AbstractCommand;
import com.github.zly2006.usm.api.event.EventSource;
import com.github.zly2006.usm.api.event.Listener;
import com.github.zly2006.usm.api.Permission;
import com.github.zly2006.usm.api.text.Text;
import com.github.zly2006.usm.api.user.User;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.nio.file.Path;
import java.util.UUID;

public interface Instance extends EventSource {
    boolean running();
    void start();
    void stop();
    @NotNull
    UUID id();
    @NotNull
    Path root();
    @NotNull
    Text getName();
    @NotNull
    AbstractCommand startCommand();
    @NotNull
    AbstractCommand stopCommand();
    @Nullable
    InstanceFolder folder();
    default boolean checkPermission(User user, String permission) {
        return checkPermission(user, Permission.valueOf(permission));
    }
    boolean checkPermission(User user, Permission permission);
    default void setPermission(User user, String permission, boolean value) {
        setPermission(user, Permission.valueOf(permission), value);
    }
    void setPermission(User user, Permission permission, boolean value);
}

File: usm/src/main/kotlin/com/github/zly2006/usm/impl/instance/Instance.kt
package com.github.zly2006.usm.impl.instance

import com.github.zly2006.usm.api.AbstractCommand
import com.github.zly2006.usm.api.Permission
import com.github.zly2006.usm.api.instance.InstanceFolder
import com.github.zly2006.usm.api.text.Text
import com.github.zly2006.usm.api.user.User
import com.github.zly2006.usm.impl.EventSource
import java.nio.file.Path
import java.util.*

class Instance(
    val name: String,
    val uuid: UUID = UUID.randomUUID(),
    val root: Path,
    val startCommand: AbstractCommand,
    val stopCommand: AbstractCommand,
): EventSource(), com.github.zly2006.usm.api.instance.Instance {
    var running = false
    var folder: InstanceFolder? = null
    override fun running() = running
    override fun start() {
        if (running) {
            return
        }
        running = true
        startCommand().execute(null)
    }

    override fun stop() {
        stopCommand().executeFuture(null).join()
        running = false
    }

    override fun id() = uuid
    override fun root() = root
    override fun getName() = Text.of(name)
    override fun startCommand() = startCommand
    override fun stopCommand() = stopCommand
    override fun folder() = folder

    override fun checkPermission(user: User?, permission: Permission?): Boolean {
        TODO("Not yet implemented")
    }

    override fun setPermission(user: User?, permission: Permission?, value: Boolean) {
        TODO("Not yet implemented")
    }
}